src/sql/src/query_model/validator/quantifier.rs

```diff
@@ -0,0 +1,407 @@
+            ValidationError::InvalidInputQuantifiers(box_except, ONE_OR_MORE_FOREACH),
+            ValidationError::InvalidInputQuantifiers(box_except, ZERO_NOT_FOREACH),
+            ValidationError::InvalidInputQuantifiers(box_intersect, ONE_OR_MORE_FOREACH),
+            ValidationError::InvalidInputQuantifiers(box_intersect, ZERO_NOT_FOREACH),
+        ]);
+        assert_eq!(errors_act, errors_exp);
+    }
+
+    /// Tests the "a Grouping box must have a single input quantifier of type Foreach" case .
+    #[test]
+    fn test_invalid_grouping() {
+        let mut model = Model::new();
+
+        let box_src = model.make_box(get_user_id(0).into());
+        let box_grouping = model.make_box(Grouping::default().into());
+        let box_dst = model.make_box(BoxType::Union);
+
+        model.make_quantifier(QuantifierType::Existential, box_src, box_grouping);
+        model.make_quantifier(QuantifierType::Foreach, box_src, box_grouping);
+        model.make_quantifier(QuantifierType::Foreach, box_grouping, box_dst);
+        model.make_quantifier(QuantifierType::Foreach, box_grouping, box_dst);
+
+        let result = QuantifierConstraintValidator::default().validate(&model);
+        assert!(result.is_err());
+
+        let errors_act: HashSet<_> = result.unwrap_err().drain(..).collect();
+        let errors_exp = HashSet::from([
+            ValidationError::InvalidInputQuantifiers(box_grouping, ZERO_NOT_FOREACH),
+            ValidationError::InvalidInputQuantifiers(box_grouping, ONE_FOREACH_INPUT_SELECT),
+            ValidationError::InvalidRangingQuantifiers(box_grouping, ONE_FOREACH_PARENT_SELECT),
+        ]);
+        assert_eq!(errors_act, errors_exp);
+    }
+
+    /// Tests constraints for quantifiers incident with [`BoxType::Select`] boxes.
+    #[test]
+    fn test_invalid_select() {
+        let mut model = Model::new();
+
+        let box_src = model.make_box(get_user_id(0).into());
+        let box_select = model.make_box(Select::default().into());
+        let box_dst = model.make_box(BoxType::Union);
+
+        model.make_quantifier(QuantifierType::PreservedForeach, box_src, box_select);
+        model.make_quantifier(QuantifierType::Foreach, box_select, box_dst);
+        model.make_quantifier(QuantifierType::Foreach, box_select, box_dst);
+
+        let result = QuantifierConstraintValidator::default().validate(&model);
+        assert!(result.is_err());
+
+        let errors_act: HashSet<_> = result.unwrap_err().drain(..).collect();
+        let errors_exp = HashSet::from([
+            ValidationError::InvalidInputQuantifiers(box_select, ONE_OR_MORE_NOT_PRES_FOREACH),
+            ValidationError::InvalidInputQuantifiers(box_select, ZERO_PRESERVED_FOREACH),
+        ]);
+        assert_eq!(errors_act, errors_exp);
+    }
+
+    /// Tests constraints for quantifiers incident with [`BoxType::OuterJoin`] boxes.
+    #[test]
+    fn test_outer_join() {
```

Comment:
We should add a follow-on issue to #8073 to add more comprehensive test cases after we have support for a DSL to test QGM. For example, ideally this section would also test that our validation system ensures the presence of at least one PreservedForeach quantifier in an OuterJoin box.

Reply:
>  For example, ideally this section would also test that our validation system ensures the presence of at least one PreservedForeach quantifier in an OuterJoin box.

This is the `ONE_OR_TWO_PRES_FOREACH` and this test case already asserts that violations of this constraint are caught with its sample input.

Reply:
What I meant was that ideally we would have additional test cases like:
* tests where we expect no error to be thrown. e.g. an outer join box with 2 preserve foreach quantifiers and nothing else..
* boxes violating constraints in multiple different ways, e.g. an empty outer join box, an outer join box with 3 preserved foreach quantifiers and nothing else.

Reply:
For the multiple different ways, if we want to test every possibility we will have exponential blowup, so at least for outer join it seems that there might be exponential number of 2*2*2=8 different cases for all possible combinations of the first 3 cases (constraint #4 is the negation of constraint #3).

What I did here is to produce a case that violates all constraints. As constraints are tested independently, this still exercises the entire code path (so code coverage won't change if we add extra cases if we can measure it). Granted, this violates the "black box" assumption that the tests should have w.r.t. the implementation...

I think it makes sense to have a basic "no errors" test and I might include one such test for each case before I merge this.

For the OuterJoin I can think an extra case that will exercise the missing constraint (maybe 2 foreach quantifiers).